shell/platform/darwin/macos/framework/Source/FlutterVSyncWaiterTest.mm

```diff
@@ -0,0 +1,189 @@
+    EXPECT_EQ(baton, size_t(1));
+  };
+
+  // First argument if the wait duration after reference vsync.
+  // Second argument is the expected delay between reference vsync and synthesized vsync.
+  test(0.005, displayLink.nominalOutputRefreshPeriod);
+  test(0.025, 2 * displayLink.nominalOutputRefreshPeriod);
+  test(0.040, 3 * displayLink.nominalOutputRefreshPeriod);
+}
+
+TEST(FlutterVSyncWaiterTest, VSyncWorks) {
+  TestDisplayLink* displayLink = [[TestDisplayLink alloc] init];
+  displayLink.nominalOutputRefreshPeriod = 1.0 / 60.0;
+  const uintptr_t kWarmUpBaton = 0xFFFFFFFF;
+
+  struct Entry {
+    CFTimeInterval timestamp;
+    CFTimeInterval targetTimestamp;
+    size_t baton;
+  };
+  __block std::vector<Entry> entries;
+
+  FlutterVSyncWaiter* waiter = [[FlutterVSyncWaiter alloc]
+      initWithDisplayLink:displayLink
+                    block:^(CFTimeInterval timestamp, CFTimeInterval targetTimestamp,
+                            uintptr_t baton) {
+                      entries.push_back({timestamp, targetTimestamp, baton});
+                      if (baton == kWarmUpBaton) {
+                        return;
+                      }
+                      EXPECT_TRUE(CACurrentMediaTime() >= timestamp - kTimerLatencyCompensation);
+                      CFRunLoopStop(CFRunLoopGetCurrent());
+                    }];
+
+  [waiter waitForVSync:kWarmUpBaton];
+
+  // Reference vsync to setup phase.
+  CFTimeInterval now = CACurrentMediaTime();
+  // CVDisplayLink callback is called one and a half frame before the target.
+  [displayLink tickWithTimestamp:now + 0.5 * displayLink.nominalOutputRefreshPeriod
+                 targetTimestamp:now + 2 * displayLink.nominalOutputRefreshPeriod];
+  EXPECT_EQ(displayLink.paused, YES);
+
+  [waiter waitForVSync:1];
+  CFRunLoopRun();
+
+  [waiter waitForVSync:2];
+  [displayLink tickWithTimestamp:now + 1.5 * displayLink.nominalOutputRefreshPeriod
+                 targetTimestamp:now + 3 * displayLink.nominalOutputRefreshPeriod];
+  CFRunLoopRun();
+
+  [waiter waitForVSync:3];
+  [displayLink tickWithTimestamp:now + 2.5 * displayLink.nominalOutputRefreshPeriod
+                 targetTimestamp:now + 4 * displayLink.nominalOutputRefreshPeriod];
+  CFRunLoopRun();
+
+  EXPECT_FALSE(displayLink.paused);
+  // Vsync without baton should pause the display link.
+  [displayLink tickWithTimestamp:now + 3.5 * displayLink.nominalOutputRefreshPeriod
+                 targetTimestamp:now + 5 * displayLink.nominalOutputRefreshPeriod];
+  // Make sure to run the timer scheduled in display link callback.
+  CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.02, NO);
+  ASSERT_TRUE(displayLink.paused);
+
+  EXPECT_EQ(entries.size(), size_t(4));
+
+  // Warm up frame should be presented as soon as possible.
+  EXPECT_TRUE(fabs(entries[0].timestamp - now) < 0.005);
+  EXPECT_TRUE(fabs(entries[0].targetTimestamp - now) < 0.005);
```

Comment:
Potentially worth extracting a named constant, but with only two cases, it's likely not really worth it. We probably have precedent for naming such a thing, ideally with a better name than `kTimingWiggleRoom`.